[자동차 경주 미션 - 1단계] 홍여준 미션 제출합니다.

programming-alpaca opened the pull request:
안녕하세요! 

이번에 모든 기능 구현 후 일괄 제출 하였으며, 어려웠던 점은 다음과 같습니다.
- 마지막에 수상자 출력하는 과정에서 이름과 쉼표를 그냥 번갈아가면서 찍으면 쉼표가 추가로 찍히는 과정이 발생했습니다.
- 그래서 if문으로 다수 수상자의 경우와 단일 수상자로 나누었으나 3항 연산자 규칙에 어긋났습니다.
- 그래서 쉼표를 추가하는 기능을 새로운 메서드로 빼서 처리하였습니다.

감사합니다!

programming-alpaca committed:
docs(Readme.md) : add 구현할 기능 목록

Add 구현할 기능 목록

programming-alpaca committed:
feat : add 기능들

추가된 기능들
- 자동차 이름 입력받기
- 시도 회수 입력받기
- 시도 회수 만큼 난수 생성 후 조건에 따라 자동차 전진 및 그대로 유지
- 매 시도가 끝나면 현 진행상황 출력
- 게임이 끝나면 자동차들의 위치 중 최댓값 찾고 최댓값과 자동차들의 위치를 비교하고 우승한 자동차 이름 출력하기

BGuga commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,7 +1,88 @@
+        }
+    }
+
+    public static void PrintWinners(List<Car>CarsList) {
```
메서드 명은 lowerCamelCase 로 작성해주세요!!

programming-alpaca replied:
넵 확인했습니다! public static void printWinners() 맞을까요?

BGuga commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,7 +1,88 @@
-        // TODO 구현 진행
+        List<Car> Cars = CreateCars();
+        PlayGame(Cars);
+        PrintWinners(Cars);
```
저는 하나의 클래스 내에서 메서드 배치를 호출되는 순서로 적고 외부에서 호출하는 메서드들을 상단에 배치합니다.😄
이 방식대로 한번 메서드들을 재배치 해보고 가독성이 더 나아졌는지 한번 비교해보시면 좋을 것 같습니다.

programming-alpaca replied:
해당 방식 이해를 잘 못했습니다! 현재 메서드 배치가 순서대로 되어 있지 않나요...? printWinners()에서 getName()을 호출하고 getMaxPosition() 에서 getPosition()을 호출하니 getMaxPosition()을 위로 상단으로 배치한다는 말씀이실까요?

BGuga replied:
표현이 정확하지 못했군요!
main 메서드가 시작점이니 제일 상단에 위치시킨 후 먼저 호출되는 순서로 배치하라는 뜻이였습니다.

BGuga commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,7 +1,88 @@
+    public static void PrintWinners(List<Car>CarsList) {
+        Integer[] positionsArray = GetPosition(CarsList);
+        int winnercount = 0;
+        int maxposition = positionsArray[positionsArray.length - 1];
```
배열 대신 Collection 을 사용하면 좋을 것 같아요 ㅎㅎ
물론 배열과 Collection 모두 각자의 장,단점이 있지만 대부분 Collection 을 선택합니다.
왜 그럴까요??

programming-alpaca replied:
Collection을 사용하게 된다면 Array에 비해서 상대적으로 유동적으로 데이터 구조를 수정할 수 있을 것 같습니다! 그리고 배열에 비해서 다룰수 있는 데이터 종류가 많아서 대부분 collection을 사용하는 것 같지만 상황에 따라 배열과 collection을 선택적으로 사용하면 될 것 같습니다!

BGuga replied:
```
그리고 배열에 비해서 다룰수 있는 데이터 종류가 많아서 대부분 collection을 사용하는 것 같지만
```
오히려 Collection은 기본형을 사용하지 못하기 때문에 다룰 수 있는 데이터는 한정적입니다.
하지만 Collection이 제공해주는 기능에 더 의미를 두어 Collection을 선호하며 배열 크기를 늘려주는 기능은 좀 더 객체지향 구조에 알맞습니다.

BGuga commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,7 +1,88 @@
+        }
+    }
+
+    public static Integer[] GetPosition(List<Car> CarsList) {
+        List<Integer> positions = new ArrayList<>();
+        for (Car car : CarsList) {
+            positions.add(car.Position());
+        }
+        Integer[] positionsArray = positions.toArray(new Integer[0]); // positions 리스트의 길이를 가진 Array 새롭게 생성 어차피 0과 positions의 길이를 비교한 후 큰 값으로 생성됨
+        Arrays.sort(positionsArray);
+        return positionsArray;
+    }
```
배열로 바꿀 필요 없이 Collections 에 선언된 메서드를 통해 Collection 정렬을 할 수 있습니다 ㅎㅎ

BGuga replied:
메서드 명은 getPoisition 인데 실질적으로 하는 작업은 자동차 중 가장 멀리 나간 값을 구하는 것이네요
```java
public static Integer getMaxPosition(List<Car> carList)
```
그러면 메서드를 이렇게 구현해줘도 될 것 같은데 다른 이유가 있으실까요??

programming-alpaca replied:
2. getPosition() 메서드에서 각 자동차들의 숫자들을 리스트로 구현하고 나중에 그 리스트 중에서 최대 숫자를 추출하려고 했으나 다시 읽어보니 굳이 2번의 절차를 걸치지 않고 그냥 바로 최대 숫자를 반환했었어도 괜찮을 것 같습니다!

programming-alpaca replied:
1. 배열로 안바꾸고 변환을 하게 된다면 다음과 같은 코드가 될까요?
`Collections.sort(positions);`

BGuga replied:
이렇게 만들 수도 있습니다!

```java
public static Integer getMaxPosition(List<Car> carList) {
        return Collections.max(carList, Comparator.comparingInt(Car::getPosition))
            .getPosition();
}
```

BGuga commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,7 +1,88 @@
+        }
+    }
+
+    public static void checkError(String[] CarsList) {
+        for (String str : CarsList) {
+            if (str.isEmpty() || str.length() >= 6) { //5자 이하여야 하나 공백 역시 허용 안됨
+                throw new IllegalArgumentException("[ERROR] 자동차의 이름은 5자 이하여야 합니다");
+            }
+        }
+    }
```
정확한 행위를 짚어내기 위해서 메서드 명을 validateName 으로 바꾸는 것은 어떨까요??

programming-alpaca replied:
훨씬더 명확하게 메서드의 행위를 알려줄 것 같습니다!

BGuga commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,7 +1,88 @@
+        PrintWinners(Cars);
     }
-}
+}
```
[다음 글](https://seongwon.dev/Git/20220303-%ED%8C%8C%EC%9D%BC%EC%9D%98_%EB%A7%88%EC%A7%80%EB%A7%89_%EA%B0%9C%ED%96%89/)을 읽어보시고 끝 줄에 개행을 추가해주세요 !!

programming-alpaca replied:
전혀 생각지 못했던 부분이네요! 좋은 글 감사합니다!

BGuga replied:
Car도 개행 추가해주세요!

BGuga commented on src/main/java/racingcar/Car.java:
```diff
@@ -8,5 +9,20 @@ public Car(String name) {
+        }
+        System.out.print("\n");
+    }
+    public String Name() {
```
Class의 필드를 가져오는 메서드는 get + 필드 명으로 작성하는게 관례입니다!

programming-alpaca replied:
넵! 알겠습니다!

programming-alpaca replied:
넵 알겠습니다!

BGuga commented on src/main/java/racingcar/Car.java:
```diff
@@ -8,5 +9,20 @@ public Car(String name) {
+            System.out.print("-");
+        }
+        System.out.print("\n");
+    }
```
메서드 사이에 공백 추가해주세요~

programming-alpaca replied:
넵! 알겠습니다!

programming-alpaca replied:
넵 알겠습니다!

BGuga commented on src/main/java/racingcar/Car.java:
```diff
@@ -8,5 +9,20 @@ public Car(String name) {
-    // 추가 기능 구현
-}
+    public void moveCar() {
+        if (Randoms.pickNumberInRange(0,9) >= 4) {
+            position++;
+        }
+        System.out.print(this.name+" : ");
+        for (int i = 0; i < position; i++){
+            System.out.print("-");
+        }
+        System.out.print("\n");
```
이 기능은 Car 안에 구현하신 이유가 있을까요??

programming-alpaca replied:
Car Class 안에 있는 private 변수는 name과 position인데, 게임을 진행하면서 이름을 수정하는 경우는 발생하지 않지만 position을 수정하는 경우는 발생합니다. 이 때, position을 변경하는 함수인 moveCar()이 Car.java에 있어야 public class의 private 변수를 class 외부에서 접근할 수 있을 것이라고 생각했습니다!

BGuga replied:
그러면
```java
//Applicatoin

if (num > 4) {
   car.move()
}
```

```java
public void move() {
        position++;
}
```
로도 구현할 수 있는데 4 이상일 경우 이동한다는 규칙을 Car에 담아낸 이유가 궁금했습니다!!
여준님이 해준 것 처럼 자동차에 규칙이 담겨있으면 객체지향적이기에 더 좋습니다.

BGuga reviewed (changes requested):
안녕하세요 여준님 수고 많으셨습니다~
피드백 반영하면서 어려운 부분 있으면 말씀 주세요 ㅎㅎ

programming-alpaca committed:
Changed Method Names (lowerCamelCase)

programming-alpaca committed:
코드 포맷 변경 : 개행 추가

코드 포맷 변경 : 메서드 이름 변경
클래스의 필드를 가져오는 메서드 이름을 get + 필드명으로 변경

코드 포맷 변경 : 메서드간 공간 추가

programming-alpaca committed:
코드 포맷 수정 : 메서드 이름 변경

메서드의 행위를 더 알기 쉽게 변경

programming-alpaca committed:
코드 포맷 수정 : 변수 이름 변경

변수를 더 확인하기 쉬워짐

코드 수정 : 메서드 이름 & 코드 수정

getPosition -> getMaxPosition()으로 변경
여러 절차를 거치지 않고 메서드에서 바로 숫자를 반환
